feat(windows): listen for DNS change events

ReactorScram opened the pull request:
```[tasklist]
- [x] Ensure whatever method we use to read the DNS servers actually works here, and doesn't have a strange memory ordering problem. If it does, read them from the registry by hand.
- [x] Graceful exit
- [x] Error handling
- [x] Clean it up and integrate it with the Tauri client
- [x] Replace `Notify` with channel of size one
- [x] Remove `Arc`
- [x] Replace `close` with panics
- [x] Remove `Pin`
- [x] Unit-test what happens if we register with RegNotify, close the handle, then modify our key
- [x] Merge with main and call `set_dns`
```

ReactorScram committed:
listen for DNS changes asynchronously on IPv4 and IPv6

ReactorScram committed:
refactor

ReactorScram committed:
tested this one manually. Looks like it picks up the DNS correctly

ReactorScram committed:
fmt

ReactorScram committed:
hook in

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into feat/windows/network-and-dns-changes

ReactorScram committed:
hook up `Session::reconnect`

ReactorScram commented on rust/gui-client/src-tauri/Cargo.toml:
```diff
@@ -43,7 +43,7 @@ tauri = { version = "1.6", features = [ "dialog", "shell-open-api", "system-tray
 tauri-utils = "1.5.3"
 thiserror = { version = "1.0", default-features = false }
-tokio = { version = "1.36.0", features = ["time"] }
+tokio = { version = "1.36.0", features = ["signal", "time"] }
```
I wanted graceful Ctrl+C for the debug subcommand. It's not used in the GUI itself, although there's no reason we couldn't.

jamilbk replied:
I think this is also why the Gateway doesn't respond to Ctrl-C? I think that causes the docker container to delay restart/shutdown

ReactorScram replied:
I had thought it was because the `init` flag wasn't set in our Dockerfile, but I can't remember if that fixed it. It seemed to help for the Client.

I've definitely seen it before on other Docker containers, you have to catch Ctrl+C and also tell Docker to forward it.

ReactorScram committed:
fmt, cross-platform, etc.

ReactorScram committed:
remove incorrect comment

jamilbk commented on rust/gui-client/src-tauri/src/client/gui.rs:
```diff
@@ -820,8 +822,18 @@ async fn run_controller(
+                if let Some(session) = controller.session.as_mut() {
+                    tracing::debug!("New DNS resolvers, calling `Session::reconnect`");
+                    // TODO: Pass DNS resolvers to connlib here
+                    session.connlib.reconnect();
```
AFAIK `reconnect()` is for network changes and `set_dns(dns_servers: Vec<IpAddr>)` is for DNS. At least that [was my assumption here](https://github.com/firezone/firezone/blob/67c828f013266e85833e8d2b2ffa07eca7f34799/swift/apple/FirezoneNetworkExtension/Adapter.swift#L284).

```suggestion
                    tracing::debug!("New DNS resolvers, calling `Session::set_dns`");
                    // TODO: Pass DNS resolvers to connlib here
                    session.connlib.set_dns([]);
```

Granted on some platforms they'll always be called at the same time (macOS, Android?)

ReactorScram replied:
You're right, I just thought `set_dns` wasn't merged yet. I'll check again

jamilbk replied:
Ah, I'm just stubbing it out. I was thinking it was going to be added for this refactor effort

ReactorScram replied:
Yeah that'll get added in #4163 https://github.com/firezone/firezone/pull/4163/files#diff-0b4003993c22f01feddc49f811f4fde0f8e4111b3c655a6104ee0564efbd5d38R84

Not sure if it makes more sense to keep `reconnect` here or get rid of it and just leave the comments

jamilbk commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+            // so he might be talking about that.
+            // If he's not, then we're invisibly tying up two worker threads. Can't help it.
+
+            // SAFETY: It's complicated.
```
😆

jamilbk commented on rust/gui-client/src-tauri/src/client/gui.rs:
```diff
@@ -820,8 +822,18 @@ async fn run_controller(
-                    tracing::info!(?have_internet);
+                    if let Some(session) = controller.session.as_mut() {
+                        tracing::debug!("Internet up/down changed, calling `Session::reconnect`");
+                        session.connlib.reconnect();
```
👍

jamilbk commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
Just so I'm following:

- We can't grab DNS-only updates from the Registry
- So we subscribe to whole-interface updates
- When we receive one, we fetch dns_servers again and send them to `Session.set_dns`

Is that about right?

Is it redundant to have the other check then that triggers only `reconnect()`? I.e. can this notify mechanism just trigger the `reconnect()` and `set_dns()` at the same time? Would that be any cleaner?

ReactorScram replied:
I'll try it again and document it, but I think this registry key does not change when we gain or lose Internet, because that's an ephemeral thing whereas Windows may store DNS in the registry so it can outlive reboots? 🤔 

And yeah I subscribed to almost anything that changes in the registry relating to any IPv4 or IPv6 interface. Being more specific would add a lot of little edge cases, worst-case resource leaks if I didn't unsubscribe correctly when a network interface was removed. This way it's exactly 2 keys, and I keep them even if Firezone signs out and back in.

ReactorScram replied:
Hm if I turn Wi-Fi off, wait a bit, and turn it back on, the DNS listener catches when I turn it on (presumably because DHCP kicks in) but not when it goes off.

So if a "Lost Internet" is useful to connlib at all, let's keep both for now.

jamilbk reviewed (commented):
Nice work! Just some clarifying questions.

Hopefully this is the deepest integration we'll need with the Registry for a while...

![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExenV4dnhlNDltbG16cTJvaHlreTI0ZjBmb3gxZ3kxZWt2eWM0NXkzciZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/pynZagVcYxVUk/giphy.gif)

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into feat/windows/network-and-dns-changes

ReactorScram committed:
Merge branch 'main' into feat/windows/network-and-dns-changes

conectado commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -188,6 +162,8 @@ pub fn check_internet() -> Result<bool> {
 pub(crate) struct Worker {
     inner: Option<WorkerInner>,
+    // I think this is `Arc` because of the worker thread, and we need the worker thread
+    // because of COM. `Notify` only has `&self` methods so it does not need `Arc` in general.
```
Why not remove the `Arc`?

ReactorScram replied:
It gets shared with another thread. But I have to replace it with an `mpsc` channel anyway per #4226, and when that happens I can remove the `Arc` because it'll be two separate tx and rx pieces.

ReactorScram replied:
#4227 removes that `Arc`

conectado commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+            Ok(that)
+        }
+
+        pub(crate) fn close(&mut self) -> Result<()> {
```
We should inline this in `drop`

ReactorScram replied:
I wanted it separate so I can handle errors if closing fails

ReactorScram replied:
I think all I do with the errors is log a `tracing::error!` so I could inline that, too. But this felt more right in terms of, "What if someone was actually using this API and it wasn't just tightly integrated with the Windows client"

conectado replied:
having to explicitly `close` goes against RAII.

I think in this case would even make sense to panic in case of errors.

conectado commented on rust/gui-client/src-tauri/src/client/gui.rs:
```diff
@@ -859,10 +871,14 @@ async fn run_controller(
         }
     }
 
+    if let Err(error) = dns_listener.close() {
```
There's no need to `close` explicitly since it's in `drop`

conectado commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+    /// Opens and returns the IPv4 and IPv6 registry keys
+    ///
+    /// `HKEY_LOCAL_MACHINE/SYSTEM/CurrentControlSet/Services/Tcpip[6]/Parameters/Interfaces`
+    fn open_network_registry_keys() -> Result<[winreg::RegKey; 2]> {
```
A using a tuple is more standard in rust

conectado commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+
+    struct Listener {
+        key: winreg::RegKey,
+        notify: std::pin::Pin<Box<Notify>>,
```
Why have `Notify` pinned?

ReactorScram replied:
I added a comment to explain it, since it's safety-related and I should have it written down: https://github.com/firezone/firezone/pull/4198/files#diff-f8b44e003aaf67d1e8ce86b898c00ae49ebb1ba2ac6d5ae40ec47fdcf64e996dR504-R507

conectado replied:
this applies here too https://github.com/firezone/firezone/pull/4198#discussion_r1533135656

conectado commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+            // SAFETY: It's complicated.
+            // The callback here can cause a lot of problems. We pin the `Notify` object.
+            // We don't use an `Arc` because `Notify` already has only `&self` methods, and
+            // the callback has no way to free an `Arc`.
```
The callback can decrement the `Arc` using `Arc::from_raw`

ReactorScram replied:
But when `drop` is called, we'll cancel the callback, which will forget the callback's cloned `Arc`, if the callback has not fired. We could check the `Receiver` to see whether we need to force the Arc to drop, but I thought it's simpler to imagine that the callback is taking a const borrow from `Listener`, and `Drop` enforces the lifetimes, and not that they're actually sharing ownership.

conectado replied:
but if we use the `arc` we don't need to unregister the callback, right?

ReactorScram replied:
I don't understand. This one is the long-lived callback registered by `RegisterWaitForSingleObject` and unregistered in `Drop` by `UnregisterWaitEx`. If I didn't unregister the callback, it would leak the `Arc` and 1 or 2 Win32 handles.

conectado commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+        }
+
+        // Be careful with this, if you register twice before the callback fires,
+        // it will probably leak something. Check the MS docs for `RegNotifyChangeKeyValue`.
```
This should be an `unsafe` function

ReactorScram replied:
Because of the constraint mentioned in that comment?
🤔 Since it's all `&mut self`, what if I enforce it at runtime with a flag `callback_armed`?

conectado replied:
I mean, it's okay to mark this as unsafe, specially since we use it only internally, the methods using it can make sure that they are enforcing the constaints.

ReactorScram replied:
I don't think it's memory-unsafe, I was thinking of this sentence from the docs https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regnotifychangekeyvalue#remarks

> Each time a process calls RegNotifyChangeKeyValue with the same set of parameters, it establishes another wait operation, creating a resource leak. Therefore, check that you are not calling RegNotifyChangeKeyValue with the same parameters until the previous wait operation has completed.

I might write a unit test for some of the registry stuff, there's no equivalent `Unregister` function, there's just closing the event handle. Even though it's only closed at the end of the process, it would be nice to know that editing the key won't break anything.

conectado commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+        // There is no reason in practice you'd ever want to call it from two
+        // threads at once.
+        //
+        // Cancel-safety: Probably. `tokio::sync::Notify` sounds cancel-safe.
```
[it's not](https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety) we should use `mpsc::channel` with buffer size 1 if we want cancellation safety.

ReactorScram replied:
Darn. I thought "lose your place in the queue" would be irrelevant if there's only one waiter.
Done

conectado commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+        key: winreg::RegKey,
+        notify: std::pin::Pin<Box<Notify>>,
+        wait_handle: Option<HANDLE>,
+        event: Option<HANDLE>,
```
Does these need to be an `Option`? If we don't expose `close` we expect to always have them.

conectado commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+            // When we call `UnregisterWaitEx` later, we wait for all callbacks to finish running
+            // before we drop the `Notify`, to avoid a dangling pointer.
+            unsafe {
+                RegisterWaitForSingleObject(
```
just to make sure I understand this `RegisterWaitForSingleObject` will signal `wait_handle`, every time `event` is signaled, even when the event that `event` is listening for changes?

ReactorScram replied:
`wait_handle` doesn't signal, it's just a handle we use to un-register the callback when dropping. I added comments here just now: https://github.com/firezone/firezone/pull/4198/files#diff-f8b44e003aaf67d1e8ce86b898c00ae49ebb1ba2ac6d5ae40ec47fdcf64e996dR505-R515

It does fire the callback every time. We only need to call `RegisterWaitForSingleObject` once, but we have to call `RegNotifyChangeKeyValue` repeatedly.

What do you mean, "the event that `event` is listening for changes"?

conectado reviewed (commented):
Looks good! Just left a few comment but nothing big

ReactorScram committed:
Merge branch 'feat/windows/network-and-dns-changes' of github.com:firezone/firezone into feat/windows/network-and-dns-changes

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into feat/windows/network-and-dns-changes

ReactorScram committed:
change array of 2 to tuple

ReactorScram committed:
fix cancellation safety

ReactorScram committed:
fix comment

ReactorScram committed:
add comments

ReactorScram committed:
fix, this should be marked thread-agnostic

ReactorScram committed:
fmt

ReactorScram committed:
fix comments

ReactorScram committed:
comment

ReactorScram committed:
todo

conectado commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -422,3 +399,240 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+        // - `Listener::new` called, pointer to `tx` passed to `RegisterWaitForSingleObject`
+        // - `Listener` and `tx` move
+        // - The callback fires, using the now-invalid previous location of `tx`
+        _tx: std::pin::Pin<Box<mpsc::Sender<()>>>,
```
I don't think this needs to be `Pin`, If `Listener` moves, the location pointed to by the `box` wouldn't change since it's heap-allocate, that's why `Box` is `unpin` and why you can still an `&mut T` even when pinned.

conectado replied:
If this wasn't boxed, `Listener` would need to be `!unpin` but I think that'd be harder to implement without much gain

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into feat/windows/network-and-dns-changes

ReactorScram committed:
replace `close` with panics

ReactorScram committed:
Remove `Pin`

ReactorScram committed:
comments

ReactorScram committed:
fmt

ReactorScram committed:
unit test

ReactorScram committed:
fmt

ReactorScram committed:
fix other platforms

ReactorScram committed:
clippy

ReactorScram committed:
clean up reg key used for testing

I don't want to delete the whole `dev.firezone.client` key in case some other test needs it

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into feat/windows/network-and-dns-changes

ReactorScram committed:
hook in `set_dns`

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into feat/windows/network-and-dns-changes

ReactorScram committed:
log resolvers

conectado reviewed (approved):
I've tested that network switching works with this branch, I'd be okay to merge :)

ReactorScram committed:
Merge branch 'main' into feat/windows/network-and-dns-changes

ReactorScram committed:
fix merge problem

ReactorScram committed:
fix comment

thomaseizinger commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -429,3 +403,337 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+
+    pub(crate) fn run_debug() -> Result<()> {
+        tracing_subscriber::fmt::init();
+        let rt = tokio::runtime::Runtime::new()?;
```
Out of curiosity, why make a new runtime here and not just make it an async function?

ReactorScram replied:
For the debug commands I don't want to start a runtime until I know which command it'll be, in case one of them needs something odd.

It's pretty rare that it matters, there's only a couple things (COM and Tauri) that could really care what thread they're on or what's happening.

thomaseizinger replied:
I see! I should have checked the code in more detailed, I was assuming that we use `async fn main`!

thomaseizinger commented on rust/gui-client/src-tauri/src/client/network_changes/windows.rs:
```diff
@@ -429,3 +403,337 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+
+        rt.block_on(async move {
+            loop {
+                tokio::select! {
+                    _r = tokio::signal::ctrl_c() => break,
+                    r = listener.notified() => r?,
+                }
```
Personally, I am not touching `tokio::select` with a 10ft pole because it has so many "interesting" features. If you only have two futures, it would suggest using `futures::future::select` instead which is trivial to understand when you look at the implementation.

If it is possible to remove the runtime above, you can even remove the `ctrl_c` listener here and just make this fn a loop of listening to be notified and compose it with `ctrl_c` at the outer layer of your app. I find listening to `ctrl_c` in `main` less surprising than several layers deep here.

YMMV :)
